File: discord_bot.py
import discord
from discord.ext import commands
import asyncio
import logging
from datetime import datetime
from config import Config

class YieldBot(commands.Bot):
    def __init__(self):
        intents = discord.Intents.default()
        # Only use non-privileged intents
        intents.guilds = True
        intents.guild_messages = True
        super().__init__(command_prefix='!', intents=intents)
        
        self.channel_id = Config.DISCORD_CHANNEL_ID
        self.logger = logging.getLogger(__name__)
        
    async def on_ready(self):
        """Called when the bot is ready."""
        self.logger.info(f'{self.user} has connected to Discord!')
        self.logger.info(f'Bot is in {len(self.guilds)} guilds')
        
        # Debug: List all available channels
        for guild in self.guilds:
            self.logger.info(f'Guild: {guild.name} (ID: {guild.id})')
            for channel in guild.text_channels:
                self.logger.info(f'  Channel: #{channel.name} (ID: {channel.id})')
        
        # Check if target channel is accessible
        target_channel = self.get_channel(self.channel_id)
        if target_channel:
            self.logger.info(f'✅ Target channel found: #{target_channel.name}')
        else:
            self.logger.error(f'❌ Target channel {self.channel_id} not found or not accessible')
        
    async def send_yield_notification(self, transaction_hash: str, amount: str, block_number: int, transaction_type: str = 'unknown', involves_rewards_manager: bool = False, involves_slvlusd: bool = False, events: list = None):
        """Send a yield distribution notification to the configured channel."""
        try:
            channel = self.get_channel(self.channel_id)
            if not channel:
                self.logger.error(f"Could not find channel with ID: {self.channel_id}")
                return False
                
            # Create title and description based on transaction type
            if transaction_type == 'yield_distribution_call':
                title = "🎉 Yield Distribution Detected!"
                description = f"RewardsManager method 0x6a761202 called - Protocol yield distribution triggered!"
                color = 0x00ff00  # Green color
            elif transaction_type == 'slvlusd_interaction':
                title = "💎 slvlUSD Interaction Detected!"
                description = "slvlUSD contract interaction related to yield"
                color = 0x0099ff  # Blue color
            else:
                title = "🔍 Yield-Related Transaction Detected!"
                description = "A yield-related transaction has been processed"
                color = 0x888888  # Gray color
            
            # Check if we have Rewarded events for enhanced notification
            rewarded_events = []
            if events:
                rewarded_events = [event for event in events if event.get('type') == 'Rewarded']
            
            if rewarded_events:
                title = "🏆 Rewarded Event Detected!"
                description = "Protocol yield distribution via Rewarded event - yield has been distributed!"
                color = 0xffd700  # Gold color
            
            # Create an embedded message for better formatting
            embed = discord.Embed(
                title=title,
                description=description,
                color=color,
                timestamp=datetime.utcnow()
            )
            
            embed.add_field(
                name="💰 Amount",
                value=f"{amount}",
                inline=True
            )
            
            embed.add_field(
                name="📦 Block Number",
                value=f"{block_number:,}",
                inline=True
            )
            
            embed.add_field(
                name="🔗 Transaction Hash",
                value=f"[View on Explorer](https://etherscan.io/tx/{transaction_hash})",
                inline=False
            )
            
            # Add details based on what contracts are involved
            details_parts = []
            if involves_rewards_manager:
                details_parts.append("🏛️ RewardsManager method 0x6a761202 executed")
            if involves_slvlusd:
                details_parts.append("💎 slvlUSD contract interaction")
            
            # Add Rewarded event details
            if rewarded_events:
                for event in rewarded_events:
                    if event.get('asset'):
                        details_parts.append(f"💰 Asset: {event['asset']}")
                    if event.get('recipient'):
                        details_parts.append(f"👤 Recipient: {event['recipient']}")
            
            if not details_parts:
                details_parts.append("Protocol yield distribution event detected")
            
            embed.add_field(
                name="ℹ️ Details",
                value=" • ".join(details_parts),
                inline=False
            )
            
            embed.set_footer(text="YieldBot • Monitoring Protocol Yield Distributions")
            
            await channel.send(embed=embed)
            self.logger.info(f"Yield notification sent to channel {self.channel_id}")
            return True
            
        except Exception as e:
            self.logger.error(f"Failed to send yield notification: {e}")
            return False
    
    async def send_simple_notification(self, message: str):
        """Send a simple text notification to the configured channel."""
        try:
            channel = self.get_channel(self.channel_id)
            if not channel:
                self.logger.error(f"Could not find channel with ID: {self.channel_id}")
                return False
                
            await channel.send(message)
            self.logger.info(f"Simple notification sent to channel {self.channel_id}")
            return True
            
        except Exception as e:
            self.logger.error(f"Failed to send simple notification: {e}")
            return False
    
    @commands.command(name='status')
    async def status_command(self, ctx):
        """Check bot status."""
        embed = discord.Embed(
            title="🤖 YieldBot Status",
            description="Bot is online and monitoring yield distributions",
            color=0x0099ff
        )
        
        embed.add_field(
            name="Monitoring",
            value="✅ Active",
            inline=True
        )
        
        embed.add_field(
            name="Chain",
            value=f"Chain ID: {Config.CHAIN_ID}",
            inline=True
        )
        
        await ctx.send(embed=embed)
    
    @commands.command(name='test')
    async def test_notification(self, ctx):
        """Send a test yield notification."""
        await self.send_yield_notification(
            transaction_hash="0x1234567890abcdef1234567890abcdef12345678",
            amount="1,000.50",
            block_number=18500000,
            transaction_type="yield_distribution_call",
            involves_rewards_manager=True,
            involves_slvlusd=False
        )
        await ctx.send("Test notification sent!")

def create_bot():
    """Create and return a YieldBot instance."""
    return YieldBot()

File: src/discord-bot.js
import { Client, GatewayIntentBits, EmbedBuilder } from 'discord.js';
import { config } from './config.js';

export class DiscordBot {
  constructor() {
    this.client = new Client({
      intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.MessageContent
      ]
    });
    
    this.channelId = config.DISCORD_CHANNEL_ID;
    this.readyResolve = null;
    this.setupEventHandlers();
    
  }
  
  setupEventHandlers() {
    this.client.once('ready', () => {
      console.log(`🤖 ${this.client.user.tag} has connected to Discord!`);
      console.log(`📊 Bot is in ${this.client.guilds.cache.size} guilds`);
      
      // Debug: List all available channels
      this.client.guilds.cache.forEach(guild => {
        console.log(`🏛️ Guild: ${guild.name} (ID: ${guild.id})`);
        guild.channels.cache
          .filter(channel => channel.type === 0) // Text channels
          .forEach(channel => {
            console.log(`   📝 Channel: #${channel.name} (ID: ${channel.id})`);
          });
      });
      
      // Check target channel
      this.client.channels.fetch(this.channelId)
        .then(channel => {
          if (channel) {
            console.log(`✅ Target channel found: #${channel.name}`);
          } else {
            console.error(`❌ Target channel ${this.channelId} not found or not accessible`);
          }
          
          // Resolve the start promise after channel verification
          if (this.readyResolve) {
            this.readyResolve();
          }
        })
        .catch(err => {
          console.error(`❌ Error fetching target channel ${this.channelId}: ${err.message}`);
          
          // Still resolve even if channel fetch fails
          if (this.readyResolve) {
            this.readyResolve();
          }
        });
    });
    
    this.client.on('messageCreate', async (message) => {
      if (message.author.bot) return;
      
      if (message.content === '!status') {
        await this.sendStatusMessage(message.channel);
      } else if (message.content === '!test') {
        await this.sendTestNotification(message.channel);
      }
    });
  }
  
  async start() {
    return new Promise((resolve, reject) => {
      // Store the resolve function so the ready handler can call it
      this.readyResolve = resolve;
      
      // Set up error handling
      this.client.once('error', (error) => {
        reject(error);
      });
      
      // Start the login process
      this.client.login(config.DISCORD_BOT_TOKEN).catch(reject);
    });
  }
  
  async stop() {
    if (this.client) {
      await this.client.destroy();
    }
  }
  
  async sendYieldNotification(yieldDistribution) {
    try {
      const channel = await this.client.channels.fetch(this.channelId).catch(() => null);
      if (!channel) {
        console.error(`❌ Could not find or access channel with ID: ${this.channelId}`);
        return false;
      }
      
      const { transactionHash, amount, blockNumber, transactionType, involvesRewardsManager, events } = yieldDistribution;
      
      // Create title and description based on transaction type
      let title, description, color;
      
      // Check for Rewarded events
      const rewardedEvents = events?.filter(event => event.type === 'Rewarded') || [];
      
      if (rewardedEvents.length > 0) {
        title = '🏆 Rewarded Event Detected!';
        description = 'SlvlUSD yield distribution ';
        color = 0xffd700; // Gold
      } else if (transactionType === 'yield_distribution_call') {
        title = '🎉 Yield Distribution Detected!';
        description = 'RewardsManager method 0x6a761202 called - Protocol yield distribution triggered!';
        color = 0x00ff00; // Green
      } else if (transactionType === 'slvlusd_interaction') {
        title = '💎 slvlUSD Interaction Detected!';
        description = 'slvlUSD contract interaction related to yield';
        color = 0x0099ff; // Blue
      } else {
        title = '🔍 Yield-Related Transaction Detected!';
        description = 'A yield-related transaction has been processed';
        color = 0x888888; // Gray
      }
      
      const embed = new EmbedBuilder()
        .setTitle(title)
        .setDescription(description)
        .setColor(color)
        .setTimestamp()
        .addFields(
          {
            name: '💰 Amount',
            value: amount || 'Unknown',
            inline: true
          },
          {
            name: '📦 Block Number',
            value: blockNumber.toLocaleString(),
            inline: true
          },
          {
            name: '🔗 Transaction Hash',
            value: `[View on Explorer](https://etherscan.io/tx/${transactionHash})`,
            inline: false
          }
        );
      
      // Add details
      const detailsParts = [];
      if (involvesRewardsManager) {
        detailsParts.push('🏛️ RewardsManager method 0x6a761202 executed');
      }
      
      // Add Rewarded event details
      rewardedEvents.forEach(event => {
        if (event.asset) {
          detailsParts.push(`💰 Asset: ${event.asset}`);
        }
        if (event.recipient) {
          detailsParts.push(`👤 Recipient: ${event.recipient}`);
        }
      });
      
      if (detailsParts.length === 0) {
        detailsParts.push('Protocol yield distribution event detected');
      }
      // Append weekly breakdown link
      detailsParts.push('For weekly breakdown: https://app.level.money/transparency?tab=apy');
      
      embed.addFields({
        name: 'ℹ️ Details',
        value: detailsParts.join(' • '),
        inline: false
      });
      
      embed.setFooter({ text: 'YieldBot • Monitoring Protocol Yield Distributions' });
      
      await channel.send({ embeds: [embed] });
      console.log(`📤 Yield notification sent to channel ${this.channelId}`);
      return true;
      
    } catch (error) {
      console.error(`❌ Failed to send yield notification: ${error.message}`);
      return false;
    }
  }
  
  async sendStatusMessage(channel) {
    const embed = new EmbedBuilder()
      .setTitle('🤖 YieldBot Status')
      .setDescription('Bot is online and monitoring yield distributions')
      .setColor(0x0099ff)
      .addFields(
        {
          name: 'Monitoring',
          value: '✅ Active',
          inline: true
        },
        {
          name: 'Chain',
          value: `Chain ID: ${config.CHAIN_ID}`,
          inline: true
        }
      )
      .setTimestamp();
    
    await channel.send({ embeds: [embed] });
  }
  
  async sendTestNotification(channel) {
    const testDistribution = {
      transactionHash: '0x1234567890abcdef1234567890abcdef12345678',
      amount: '31,338.00 USDC',
      blockNumber: 18500000,
      transactionType: 'yield_distribution_call',
      involvesRewardsManager: true,
      events: [{
        type: 'Rewarded',
        asset: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
        recipient: '0xdf95bb71581b224bd42eb19ceaff5e92816e181e'
      }]
    };
    
    await this.sendYieldNotification(testDistribution);
    await channel.send('✅ Test notification sent!');
  }
}
